### PR TITLE
Using Marshmallow for API input/output schema and validation

### PR DESCRIPTION
**Pull request by thinkharderdev**
This PR does a couple things:

@dgoldstein1 

1. Use Marshmallow to specify our API request/responses. This has a couple of advantages:
- We can declaratively specify API inputs/outputs along with validations in one place for consistency
- We can use the marshmallow apispec extension to auto-generate swagger specs from our schema definitions
- In cases of straightforward CRUD operations, we can use the DB entity classes to auto-generate the API request schema (see property_handler for an example)

2. Auto-generate Swagger Specs from block comments co-located with the handler methods. You can also specify request/response payloads as references to Marshmallow schemas and the swagger schema components will be auto-generated. This should be less error-prone than manually updating a giant yaml file. 

Still not done, need to finsish applying the changes. I also think we can probably take care of the method-level boilerplate for request validation so we don't have to repeat the same except block in every method.

### CODE REVIEW

File: backend/src/app/api.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+      if len(value) < 6:
+          raise ValidationError('Password must be at least 6 characters')
+housinghub_spec.components.schema("RegisterRequest", schema=RegisterRequest)
+
+class AuthResponse(Schema):
+  jwt = fields.String()
+housinghub_spec.components.schema("AuthResponse", schema=AuthResponse)
+
+class ErrorResponse(Schema):
+  message = fields.String()
```
Comment (dgoldstein1) [inline backend/src/app/api.py:34]: this should be `error`

e.g.

```json
{
  "code": 401, 
  "error": "Login invalid"
}
```

File: backend/src/auth_handlers.py [hunk 2]
```diff
@@ -1,57 +1,130 @@
...
+                {'jwt': app.services.auth_service().encode_jwt(user)})
+        return jsonify(code=401, error='Login invalid'), 401
+    except ValidationError as err:
+        app.logger.error(f'Invalid request ${err.messages}')
+        return jsonify(code=400, error=err.messages), 400
+    except:
+        app.logger.error(
+            f'Unexpected error registering new user: ${traceback.format_exc()}'
+        )
+        return jsonify(code=500, error='internal error'), 500
```
Comment (dgoldstein1) [inline backend/src/auth_handlers.py:106]: would be nice to have a test for this

File: backend/src/property_handlers.py [hunk 3]
```diff
@@ -21,11 +26,44 @@ def get_property():
...
+        payload = AddPropertyRequest().load(request.get_json())
+        _property = app.services.property_service().add_property(payload)
+        return jsonify(PropertyResponse().dump(_property))
+    except ValidationError as err:
+        app.logger.error(f'Invalid request ${err.messages}')
+        return jsonify(err.messages), 400
+    except:
+        app.logger.error(
+            f'Unexpected error adding property: ${traceback.format_exc()}')
+        return jsonify(code=500, error='internal error'), 500
```
Comment (dgoldstein1) [inline backend/src/property_handlers.py:66]: we should have a test here

Comment (thinkharderdev) [reply backend/src/property_handlers.py:66]: Working on something for testing, so I left this one out intentionally for now. Want to use https://faker.readthedocs.io/en/master/ to generate instances of our schema for unit tests so we don't have to manually create stuff.

File: backend/src/auth_handlers.py (hunk 2 above)
Comment (thinkharderdev) [reply backend/src/auth_handlers.py:106]: Done

**Comment by thinkharderdev** [comment]
> love it! Just a few basic comments:
> 
> * I am trying to register a new user:
> 
> ```shell
>  $ curl -XPOST -d '{"is_admin" : true, "password" : "password", "role" : "navigator", "username" : "user@user.com" }' http://localhost:5000/auth/register
> {
>   "code": 400,
>   "error": {
>     "_schema": [
>       "Invalid input type."
>     ]
>   }
> }
> ```
> 
> but, a) not sure where this error is coming from(?) b) is being returned in invalid schema (`error` should just be `string)
> 
> * add a few more tests in places we should have covered
> * I pushed up a commit which deletes the dockerfile line. As this causes an error on build
> 
> ```dockerfile
> # generate documentation
> RUN python api/swagger-yaml-to-html.py < api/swagger.yml > api/index.html
> ```

The error is because request.get_json() returns None if the Content-Type is not set to application/json (or one of the other mime types for JSON). I updated the curl requests in the README

**Pull request merged**